src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4117,220 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
+    insGroup*      previousTruncatedGroup = nullptr;
+#if DEBUG
+    UNATIVE_OFFSET previousJumpIgNum  = (UNATIVE_OFFSET)-1;
+    unsigned int   previousJumpInsNum = -1;
+#endif
+
+    while (jmp)
+    {
+        // if the jump is unconditional then it is a candidate
+        if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && jmp->idjIsJmpAlways)
+        {
+            insGroup* jmpGroup = jmp->idjIG;
+#if DEBUG
+            assert((jmpGroup->igFlags & IGF_HAS_ALIGN) == 0);
+            assert(jmpGroup->igNum > previousJumpIgNum || previousJumpIgNum == (UNATIVE_OFFSET)-1 ||
+                   (jmpGroup->igNum == previousJumpIgNum && jmp->idDebugOnlyInfo()->idNum > previousJumpInsNum));
+            previousJumpIgNum  = jmpGroup->igNum;
+            previousJumpInsNum = jmp->idDebugOnlyInfo()->idNum;
+#endif
+            // target group is not bound yet so use the cookie to fetch it
+            insGroup* targetGroup = (insGroup*)emitCodeGetCookie(jmp->idAddr()->iiaBBlabel);
+
+            if (targetGroup != nullptr && jmpGroup->igNext == targetGroup)
+            {
+                unsigned instructionCount = jmpGroup->igInsCnt;
+
+                if (instructionCount > 1)
+                {
+                    instrDesc* instructionDescriptor = nullptr;
+
+                    {
+                        BYTE* dataPtr = jmpGroup->igData;
+                        while (instructionCount > 0)
+                        {
+                            instructionDescriptor = (instrDesc*)dataPtr;
+                            dataPtr += emitSizeOfInsDsc(instructionDescriptor);
+                            instructionCount -= 1;
+                        };
+                        dataPtr = nullptr;
+                    }
+
+                    assert(instructionDescriptor != nullptr);
+                    // instructionDescriptor is now the last instruction in the group
+
+                    if (instructionDescriptor->idIns() == jmp->idIns() && jmp == instructionDescriptor)
+                    {
+                        // the last instruction in the group is the jmp we're looking for
+                        // and it jumps to the next instruction group so we don't need it
+
+                        JITDUMP("Removing unconditional jump IN%04x from the last instruction in "
+                                "IG%02u to the next IG IG%02u label %s\n",
+                                jmp->idDebugOnlyInfo()->idNum, jmpGroup->igNum, targetGroup->igNum,
+                                emitLabelString(targetGroup));
+
+                        // if a previous group was truncated iterate from the previous group to the
+                        // current group and ajust the offsets by the total removed byte count so far
+                        if (previousTruncatedGroup)
+                        {
+                            for (insGroup* offsetGroup                        = previousTruncatedGroup->igNext;
```

Comment:
Simplified code would to do this other way round. Starting from `jmpGroup`, adjust the `igOffs` of all IGs until we reach the next `jmpGroup`, if one exists or `emitIGlast`. That way, you won't have a separate loop below to adjust remaining IGs until the end.

Reply:
I'm not sure I agree that it would be simpler that way. It's also worth remembering that the majority of methods are going to have 1 replacement so the inner update loop won't be used very often, the outer simpler loop is the main adjustment site and it has less conditional logic.

Reply:
Having similar logic at multiple places is source of bugs and hard from maintainability point of view. We might patch one version and forget to patch another one. With having single loop, you won't have to track `previousTruncatedGroup` too.

Reply:
There isn't a way to do this which ends up with simpler code. 

To iterate into the future inside the loop requires looping through the emitJumpList list to find the next group which is going to be altered (which means duplication of all the loop condition checks) and then using a second loop to run from the current to that group or end. Any way you write this obscures the intent of the exiting loop behind the mess you get from trying to be clever to avoid a small loop. It's technically possible to do it but it isn't a win to do so and it's not something I think is worth doing without clear profile data showing that it is an improvement worth the loss of readability.

Reply:
> requires looping through the emitJumpList list to find the next group which is going to be altered

Not sure if I follow. You will have to adjust the offset for all IGs until the group that has the next jump instruction. Why can't you simply do the following outside your if-check `if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && jmp->idjIsJmpAlways)` and after you update `previousJmp` and `jmp`? This will take care of updating the offsets of IGs while you are eliminating the `jmp` instruction as well as all the IGs that follows the last `jmp` instruction.

```c++

previousJmp = jmp;
jmp         = jmp->idjNext;

insGroup* adjOffIG = previousJmp->idjIG;
insGroup* adjOffUptoIG = jmp != nullptr ? jmp->idjIG : emitIGlast;
while ((adjOffIG != nullptr) && (adjOffIG->igNum <= adjOffUptoIG->igNum))
{
    JITDUMP("Adjusted offset of %s from %04X to %04X\n", emitLabelString(adjOffIG), adjOffIG->igOffs,
            (adjOffIG->igOffs - totalRemovedSize));
    adjOffIG->igOffs -= totalRemovedSize;
    adjOffIG = adjOffIG->igNext;
}
```

Reply:
I've just tried that. To have the adjustment code in one place the `continue` inside the loop has to change to a goto and this new block gets a label. With that done there is a possible null ref of previousJmp because if the first jmp is an always there is no previous. This new code also starts too soon:

```
0 prologs, 1 epilogs, 0 funclet prologs, 0 funclet epilogs
*************** After prolog / epilog generation
G_M46441_IG01:        ; func=00, offs=000000H, size=0033H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
G_M46441_IG02:        ; offs=000033H, size=000DH, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB02 [0006], byref
G_M46441_IG03:        ; offs=000040H, size=0005H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB03 [0007], byref
G_M46441_IG04:        ; offs=000045H, size=001EH, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB05 [0001], byref
G_M46441_IG05:        ; offs=000063H, size=004EH, gcrefRegs=00000002 {rcx}, byrefRegs=00000000 {}, BB05 [0001], BB06 [0002], byref
G_M46441_IG06:        ; offs=0000B1H, size=0018H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB07 [0003], byref
G_M46441_IG07:        ; offs=0000C9H, size=0004H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB08 [0004], byref
G_M46441_IG08:        ; offs=0000CDH, size=0007H, epilog, nogc, extend
*************** In emitRemoveJumpToNextInst()
Emitter Jump List:
IG02 IN0002  je -> IG04 L_M46441_BB04
IG04 IN000a  jb -> IG05 L_M46441_BB09
IG05 IN001b  je -> IG06 L_M46441_BB07
IG05 IN001f jmp -> IG07 L_M46441_BB08 ; BB_ALWAYS
IG06 IN0025 jmp -> IG07 L_M46441_BB08 ; BB_ALWAYS
  total jump count: 5
IN001f from the last instruction in IG05 to does not jump to the next group, keeping.
Removing unconditional jump IN0025 from the last instruction in IG06 to the next IG IG07 label G_M46441_IG07
Adjusted offset of G_M46441_IG05 from 0063 to 005E
Adjusted offset of G_M46441_IG06 from 00B1 to 00AC
Adjusted offset of G_M46441_IG07 from 00C9 to 00C4
Adjusted offset of G_M46441_IG08 from 00CD to 00C8
Emitter Jump List:
IG02 IN0002  je -> IG04 L_M46441_BB04
IG04 IN000a  jb -> IG05 L_M46441_BB09
IG05 IN001b  je -> IG06 L_M46441_BB07
IG05 IN001f jmp -> IG07 L_M46441_BB08 ; BB_ALWAYS
  total jump count: 4
emitRemoveJumpToNextInst removed 1 jumps and 5 bytes
*************** In emitJumpDistBind()
Emitter Jump List:
IG02 IN0002  je -> IG04 L_M46441_BB04
IG04 IN000a  jb -> IG05 L_M46441_BB09
IG05 IN001b  je -> IG06 L_M46441_BB07
IG05 IN001f jmp -> IG07 L_M46441_BB08 ; BB_ALWAYS
  total jump count: 4
IG05 has offset 0000005E, expected 00000063

Assert failure(PID 25836 [0x000064ec], Thread: 16744 [0x4168]): Assertion failed '!"bad block offset"' in 'System.Runtime.CompilerServices.CastHelpers:LdelemaRef(System.Array,int,long):byref' during 'Generate code' (IL size 53; hash 0xd1234a96; MinOpts)

    File: E:\Programming\csharp7\runtime\src\coreclr\jit\emit.cpp Line: 1603
    Image: E:\Programming\csharp7\runtime\artifacts\bin\coreclr\windows.x64.Debug\corerun.exe
```

I can make this work and I did something pretty similar over the weekend but it's really fiddly and makes the code far harder to understand and the flow harder to follow. If it's just the repetition that bothers you I could try to extract the offset updating into a new function.

Reply:
> This new code also starts too soon

Not sure what you mean. From the assert, looks like the `igSize` is not adjusted properly? Can you elaborate why is that failure coming? 

Can you try https://github.com/kunalspathak/runtime/commit/c080feb7d48e01c4bd15d4f5035e660df5719a35 and see if that solves the problems?